crates/bevy_ecs/src/query/state.rs

```diff
@@ -506,6 +510,142 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
+        }
+    }
+
+    /// Returns a shared reference to the component `T` of the given [`Entity`].
+    ///
+    /// # Panics
+    ///
+    /// If given a nonexisting entity or mismatched component, this will panic.
+    #[inline]
+    pub(crate) fn component<'w, 's, 'r, T: Component>(
+        &'s self,
+        world: UnsafeWorldCell<'w>,
+        entity: Entity,
+    ) -> &'r T
+    where
+        'w: 'r,
+    {
+        match self.get_component(world, entity) {
+            Ok(component) => component,
+            Err(error) => {
+                panic!(
+                    "Cannot get component `{:?}` from {entity:?}: {error}",
+                    TypeId::of::<T>()
+                )
+            }
+        }
+    }
+
+    /// Returns a mutable reference to the component `T` of the given entity.
+    ///
+    /// In case of a nonexisting entity or mismatched component, a [`QueryComponentError`] is returned instead.
+    ///
+    /// # Safety
+    ///
+    /// This function makes it possible to violate Rust's aliasing guarantees.
+    /// You must make sure this call does not result in multiple mutable references to the same component.
+    #[inline]
+    pub unsafe fn get_component_unchecked_mut<'w, 's, 'r, T: Component>(
+        &'s self,
+        world: UnsafeWorldCell<'w>,
+        entity: Entity,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Result<Mut<'r, T>, QueryComponentError>
+    where
+        'w: 'r,
+    {
+        let entity_ref = world
+            .get_entity(entity)
+            .ok_or(QueryComponentError::NoSuchEntity)?;
+        let component_id = world
+            .components()
+            .get_id(TypeId::of::<T>())
+            .ok_or(QueryComponentError::MissingComponent)?;
+        let archetype_component = entity_ref
+            .archetype()
+            .get_archetype_component_id(component_id)
+            .ok_or(QueryComponentError::MissingComponent)?;
+        if self
+            .archetype_component_access
+            .has_write(archetype_component)
+        {
+            // SAFETY: It is the responsibility of the caller to ensure it is sound to get a
+            // mutable reference to this entity's component `T`.
+            let result = unsafe { entity_ref.get_mut_using_ticks::<T>(last_run, this_run) };
+
+            result.ok_or(QueryComponentError::MissingComponent)
+        } else {
+            Err(QueryComponentError::MissingWriteAccess)
+        }
+    }
+
+    /// Returns a mutable reference to the component `T` of the given entity.
+    ///
+    /// # Panics
+    ///
+    /// Panics in case of a nonexisting entity or mismatched component.
+    ///
+    /// # Safety
+    ///
+    /// This function makes it possible to violate Rust's aliasing guarantees.
+    /// You must make sure this call does not result in multiple mutable references to the same component.
+    #[inline]
+    pub(crate) unsafe fn component_unchecked_mut<'w, 's, 'r, T: Component>(
```

Comment:
One last thing -- I don't think this method needs to exist anymore. It's private and only called in one place. Leaving it here is just begging someone to make it `pub` in the future, which I don't think we should do for the same reasons I talked about for the `Query` version of this method.

Otherwise, this PR looks good to me.

Reply:
I think I disagree on that reasoning, because `many_read_only_manual`, `single`, `single_mut`, `many_unchecked_manual`, etc. all also fit that description. The downside to that change as well is the error handling in those functions then just gets shuffled back into `Query`, which I think is contrary to the goal of this PR, which is to move implementation details out of `Query` and into `QueryState`.

If you still think it should be removed I'm happy to make that change, but I think it's worth pointing out that possible inconsistency.

Reply:
It's fine to add panicking variants for `single{_mut}` and `component{_mut}` since those are safe functions. I'd even argue that they should be made public. Good point about `many_*` though -- I'd also argue that they don't need to exist, since they're private, and each only gets called once. There's not much gain to adding more unsafe functions like this IMO.

Ultimately though it's up to you. I'm just sharing my opinion :)

Reply:
And thanks for sharing! I've removed `component_unchecked_mut`, `many_read_only_manual`, and `many_unchecked_manual` as they were only called by `Query` in a single location, and were all `unsafe`. There are possibly more `unsafe` functions in `QueryState` that could be marked as `safe`, but I'm not familiar enough with Bevy or `unsafe` Rust to make that call.